### ISSUE TITLE
Improved ability to copy VSIX to output directory

### ISSUE THREAD

**Issue by sharwell** [josetr/VsixTesting#9]
It's my understanding that copying the VSIX to the integration test output directory is required for correct deployment in test scenarios. Originally I did this:

```xml
  <ItemGroup>
    <ProjectReference Include="..\Tvl.VisualStudio.MouseFastScroll\Tvl.VisualStudio.MouseFastScroll.csproj" />
  </ItemGroup>

  <ItemGroup>
    <None Include="..\Tvl.VisualStudio.MouseFastScroll\bin\$(Configuration)\net45\Tvl.VisualStudio.MouseFastScroll.vsix" CopyToOutputDirectory="PreserveNewest" />
  </ItemGroup>
```

I believe the experience would be improved if I could write this instead:

```xml
  <ItemGroup>
    <ProjectReference Include="..\Tvl.VisualStudio.MouseFastScroll\Tvl.VisualStudio.MouseFastScroll.csproj" CopyVsix="true" />
  </ItemGroup>
```

We could provide this functionality if the following was added to a **.targets** file in the NuGet package for VsixTesting.Xunit:

```xml
  <Target Name="PrepareVsixProjectReferences"
          BeforeTargets="ResolveProjectReferences"
          DependsOnTargets="PrepareProjectReferences">
    <MSBuild
        Projects="@(_MSBuildProjectReferenceExistent)"
        Targets="CreateVsixContainer;VSIXContainerProjectOutputGroup"
        BuildInParallel="$(BuildInParallel)"
        Properties="%(_MSBuildProjectReferenceExistent.SetConfiguration); %(_MSBuildProjectReferenceExistent.SetPlatform); %(_MSBuildProjectReferenceExistent.SetTargetFramework); CreateVsixContainer=true"
        Condition="'%(_MSBuildProjectReferenceExistent.CopyVsix)' == 'true'"
        ContinueOnError="!$(BuildingProject)"
        RemoveProperties="%(_MSBuildProjectReferenceExistent.GlobalPropertiesToRemove)">

      <Output TaskParameter="TargetOutputs" ItemName="_ProjectReferenceVsixOutputs" />
    </MSBuild>

    <ItemGroup>
      <None Include="@(_ProjectReferenceVsixOutputs)" Condition="Exists('%(Identity)')" CopyToOutputDirectory="PreserveNewest" />
    </ItemGroup>
  </Target>
```

**Comment by josetr** [comment]
Is there any particular reason why the default shouldn't be `true`?
`Condition="'%(_MSBuildProjectReferenceExistent.CopyVsix)' != 'false'"`

By the way, this target only works when `ExtensionsDirectory` is empty.

https://github.com/josetr/VsixTesting/blob/6df53de82de56ad83016e4da38c1f93d02a5ae9b/src/VsixTesting/ITestSettings.cs#L46

**Comment by josetr** [comment]
> The CreateVsixContainer target will not exist for project references that are not VSIX projects, so presumably it would cause the build to fail.

* Use `SkipNonExistentTargets` but it's only available since 15.5.x
* Inject empty targets using `CustomBeforeMicrosoftCommonTargets` but it's a bit hacky and wont work if you have a custom project that doesn't import `Microsoft.Common.Props`.
* Check if the target exists before executing it using MSBuild API.

Warning: I'm no MSBuild expert so this neeeds review.

Notes:
* Running task `AddCopyVsixToProjectReferences` before `ResolveProjectReferences` + running



```csproj
<Target Name="PrepareVsixProjectReferences" AfterTargets="ResolveReferences" DependsOnTargets="ResolveReferences">
    <AddCopyVsixToProjectReferences Projects="@(_MSBuildProjectReferenceExistent)" Targets="CreateVsixContainer;VSIXContainerProjectOutputGroup">
      <Output ItemName="_MSBuildProjectReferenceExistentWithCopyVsix" TaskParameter="Result" />
    </AddCopyVsixToProjectReferences>

    <MSBuild Projects="@(_MSBuildProjectReferenceExistentWithCopyVsix)" 
             Targets="VSIXContainerProjectOutputGroup" 
             BuildInParallel="$(BuildInParallel)" 
             Properties="%(_MSBuildProjectReferenceExistentWithCopyVsix.SetConfiguration); %(_MSBuildProjectReferenceExistentWithCopyVsix.SetPlatform); %(_MSBuildProjectReferenceExistentWithCopyVsix.SetTargetFramework);" Condition="'%(_MSBuildProjectReferenceExistentWithCopyVsix.CopyVsix)' != 'false'" 
             ContinueOnError="$(ContinueOnError)" 
             RemoveProperties="%(_MSBuildProjectReferenceExistentWithCopyVsix.GlobalPropertiesToRemove)">
      <Output TaskParameter="TargetOutputs" ItemName="_ProjectReferenceVsixOutputs" />
    </MSBuild>

    <ItemGroup>
      <None Include="@(_ProjectReferenceVsixOutputs)" Condition="Exists('%(Identity)')" CopyToOutputDirectory="PreserveNewest" />
    </ItemGroup>
  </Target>

  <UsingTask TaskName="AddCopyVsixToProjectReferences" TaskFactory="CodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll">
    <ParameterGroup>
      <Projects ParameterType="Microsoft.Build.Framework.ITaskItem[]" Required="true" />
      <Targets ParameterType="System.String[]" Required="true" />      
      <Properties ParameterType="System.String[]" />
      <RemoveProperties ParameterType="System.String" />
      <Result ParameterType="Microsoft.Build.Framework.ITaskItem[]" Output="true" />
    </ParameterGroup>
    <Task>
      <Reference Include="System.Xml" />
      <Reference Include="Microsoft.Build" />
      <Reference Include="Microsoft.Build.Framework" />
      <Using Namespace="Microsoft.Build.Evaluation" />
      <Using Namespace="Microsoft.Build.Execution" />
      <Using Namespace="Microsoft.Build.Utilities" />
      <Using Namespace="Microsoft.Build.Framework" />
      <Code Type="Fragment" Language="cs">
        <![CDATA[        
           foreach (var Project in Projects)
           {              
              var collection = new ProjectCollection();
              var project = collection.LoadProject(Project.ItemSpec);
              var isVsixProject = Targets.All(targetName => project.Targets.Any(target => target.Key == targetName));
              project.ProjectCollection.UnloadProject(project);
              
              if (!isVsixProject)
                Project.SetMetadata("CopyVsix", "false");
              else {
                Project.SetMetadata("CopyVsix", "true");
              }
           }
           
           Result = Projects;
      ]]>
      </Code>
    </Task>
  </UsingTask>
```
